[Ref guide] - Dojo Resources and data middleware (#722)
dojo/framework: docs/en/resource/introduction.md

# Resource

Dojo **resource** provides a consistent means to make widgets data aware.

> A Dojo resource is initialised with a `DateTemplate` that describes how to read data and enables you to create a single store of data which can be passed into multiple widgets. It allows data to be cached, transformed and filtered to suit the needs of the widget using it. Coupled with the data middleware, resources allow consistent, source agnostic data access for your widgets without them being aware of the underlying data fetching implementation or the raw data format being used as these are abstracted away by both the template read mechanism and the resource.

Reviewer: `DataTemplate`?

| ----------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Single data source**                    | Resources allows you to create a single resource for a given template and share it with multiple widgets via the data middleware.                                            |
| **Data transforms**                       | A data tranform mechanism allows you to specify the data form that your widget requires and transforms the data transparently                                                |
| **Support for async and sync data reads** | The resource template can read data in any way it sees fit, the data middleware uses a reactive API to invalidate widgets when data becomes available.                       |
| **Consistent Resource Options**           | Resource options objects are passed to all api functions ensuring that all api functions are pure and provide only the data that was requested                               |

> | **Sharable Resource Options**             | Resource options can be shared between widgets via the data middleware. This allows multiple widgets to react to changes to the resource options, ie. query, page number etc |

Reviewer 2: ```suggestion
| **Sharable Resource Options**             | Resource options can be shared between widgets via the data middleware, allowing multiple widgets to react to any changes such as modifying a query, moving to a different page number, etc. |
```

# Basic Usage

Dojo resource is comprised of a resource which you create to get and read data and a data middleware which you use within your widgets to access the resource.

> 

Reviewer 3: I think it would be useful to see the definition for a `DataTemplate` for an overview of what fields are available/optional and what they're types are. Obviously, this is information that could be found, but I reckon it's small enough to reference here.

Author: It's covered in the next paragraph, do you think that's sufficient? I didnt want to go over the template in the very first section.

## Creating a resource

A basic resource requires only a `DataTemplate` to be created which specifies a `read` function.

> main.ts

```ts
import { DataTemplate, createResource } from '@dojo/framework/core/resource';

> import { fetcher } from './personfetcher';

Reviewer: Do you think we should inline the fetcher? otherwise it's a bit opaque to the reader to fully understand what a fetcher actually is. I don't think we need to go into detail as that can happen in the supplemental section.

const template: DataTemplate = {
	read: fetcher
};

const resource = createResource(template);
```

> When a call is made to the resource api which requires data to be read, the read function will be called with the readoptions, a put function and a get function. These two helper functions can be used to side load data or to read existing data from the resource.

Reviewer 4: Should this section potentially link to or reference the supplemental doc where more details on these helper functions is provided?

> When a call is made to the resource api which requires data to be read, the read function will be called with the readoptions, a put function and a get function. These two helper functions can be used to side load data or to read existing data from the resource. More information on this can be found in the supplemental docs.

Reviewer 5: Might be good to link to the supplemental page where it is referenced.

> When a call is made to the resource api which requires data to be read, the read function will be called with the readoptions, a put function and a get function. These two helper functions can be used to side load data or to read existing data from the resource. More information on this can be found in the supplemental docs.

Reviewer: should "readoptions" either be two words or code?

Reviewer: Also should put and get be in code guards?

> ## Transforming Data

Reviewer: Interesting do we think that transforming data is a introduction topic?

Author: It's required when a widget types it's data middleware; so I felt it was

Resources provides a `createTransformer` function which can be used to create a type safe transform object. In order to correctly type the transform source values, a generic must be passed to the `DataTemplate` to type the `read` return type. When used with the `createTransformer` function, this ensures that only return values from the read function can be used as source values.

> main.ts

```ts
import { DataTemplate, createResource, createTransformer } from '@dojo/framework/core/resource';
import { fetcher } from './personfetcher';

const template: DataTemplate<{ firstName: string; lastName: string }> = {
	read: fetcher
};

const transformer = createTransformer(template, {
	value: ['firstName'],
	label: ['firstName', 'lastName']
});
```

The transformer can be passed to a widget which uses the data middleware to ensure that the data it receives from the resource is correct.

> ## Passing a Resource to a Widget

Reviewer: Should this come before showing the use of the data middleware? Seems like this would flow better being after the next section?
